"future should be done" assertion in HashBuilderOperator

Some queries are failing with this stack trace:

```
Caused by: com.google.common.base.VerifyException: future should be done
	at com.google.common.base.Verify.verify(Verify.java:123)
	at com.facebook.presto.memory.DefaultQueryContext.tryUpdateUserMemory(DefaultQueryContext.java:199)
	at com.facebook.presto.memory.DefaultQueryContext$QueryMemoryReservationHandler.tryReserveMemory(DefaultQueryContext.java:314)
	at com.facebook.presto.memory.context.RootAggregatedMemoryContext.tryUpdateBytes(RootAggregatedMemoryContext.java:49)
	at com.facebook.presto.memory.context.ChildAggregatedMemoryContext.tryUpdateBytes(ChildAggregatedMemoryContext.java:46)
	at com.facebook.presto.memory.context.ChildAggregatedMemoryContext.tryUpdateBytes(ChildAggregatedMemoryContext.java:46)
	at com.facebook.presto.memory.context.ChildAggregatedMemoryContext.tryUpdateBytes(ChildAggregatedMemoryContext.java:46)
	at com.facebook.presto.memory.context.ChildAggregatedMemoryContext.tryUpdateBytes(ChildAggregatedMemoryContext.java:46)
	at com.facebook.presto.memory.context.SimpleLocalMemoryContext.trySetBytes(SimpleLocalMemoryContext.java:77)
	at com.facebook.presto.operator.OperatorContext$InternalLocalMemoryContext.trySetBytes(OperatorContext.java:599)
	at com.facebook.presto.operator.HashBuilderOperator.updateIndex(HashBuilderOperator.java:344)
	at com.facebook.presto.operator.HashBuilderOperator.addInput(HashBuilderOperator.java:333)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:384)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:283)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:675)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:276)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1065)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:483)
	at com.facebook.presto.$gen.Presto_0_214_84_gad05dcb____20181206_001801_1.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

--- a/presto-main/src/main/java/com/facebook/presto/memory/DefaultQueryContext.java
+++ b/presto-main/src/main/java/com/facebook/presto/memory/DefaultQueryContext.java
@@ -196,7 +196,11 @@ public class DefaultQueryContext
     {
         if (delta <= 0) {
             ListenableFuture<?> future = updateUserMemory(allocationTag, delta);
-            verify(future.isDone(), "future should be done");
+            // When delta == 0 and the pool is full the future can still not be done,
+            // but, for negative deltas it must always be done.
+            if (delta < 0) {
+                verify(future.isDone(), "future should be done");
+            }
             return true;
         }
         if (queryMemoryContext.getUserMemory() + delta > maxUserMemory) {

--- a/presto-main/src/main/java/com/facebook/presto/memory/LegacyQueryContext.java
+++ b/presto-main/src/main/java/com/facebook/presto/memory/LegacyQueryContext.java
@@ -164,7 +164,11 @@ public class LegacyQueryContext
     {
         if (delta <= 0) {
             ListenableFuture<?> future = updateUserMemory(allocationTag, delta);
-            verify(future.isDone(), "future should be done");
+            // When delta == 0 and the pool is full the future can still not be done,
+            // but, for negative deltas it must always be done.
+            if (delta < 0) {
+                verify(future.isDone(), "future should be done");
+            }
             return true;
         }
         if (queryMemoryContext.getUserMemory() + delta > maxMemory) {

--- a/presto-main/src/test/java/com/facebook/presto/memory/TestMemoryTracking.java
+++ b/presto-main/src/test/java/com/facebook/presto/memory/TestMemoryTracking.java
@@ -329,6 +329,16 @@ public class TestMemoryTracking
     }
 
     @Test
+    public void testTrySetZeroBytesFullPool()
+    {
+        LocalMemoryContext localMemoryContext = operatorContext.localUserMemoryContext();
+        // fill up the pool
+        memoryPool.reserve(new QueryId("test_query"), "test", memoryPool.getFreeBytes());
+        // try to reserve 0 bytes in the full pool
+        assertTrue(localMemoryContext.trySetBytes(localMemoryContext.getBytes()));
+    }
+
+    @Test
     public void testDestroy()
     {
         LocalMemoryContext newLocalSystemMemoryContext = operatorContext.newLocalSystemMemoryContext("test");

